### ISSUE TITLE
Some issues with nativeGen

### ISSUE THREAD

**Issue by hughsando** [HaxeFoundation/hxcpp#754]

**Comment by hughsando** [comment]
The basic issue here is using a native function as a haxe closure.  Since nativeGen now tries to be as native as possible, this also means removing the dynamic glue, which is used when creating closures.
I guess the solution is to replace this with something like `function () mSampleInterface.foo()`, or throw an error at compile time.

**Comment by sorka95032** [comment]
We rarely pass functions in as closures. The place where we're doing it in the example I gave you may be the only place we do it with an @nativeGen interface.

**Comment by Simn** [comment]
I'm testing this with latest hxcpp on the Haxe 3.4.8 branch using this simplified code:

```haxe
@:nativeGen
interface SampleInterface
{
   function foo() : Void;
}

class Main implements SampleInterface
{
    private static var gMain:Main;
    private var vSampleInterface : Array<SampleInterface>;

    public static function main()
    {
		var vSampleInterface = new Array<SampleInterface>();
		vSampleInterface.push(new Main());
    }

    public function foo()
    {
       trace("called!");
    }

	function new() { }
}
```

The program crashes:

![devenv_2018-12-21_20-43-48](https://user-images.githubusercontent.com/634365/50360735-52c32c80-0561-11e9-8111-1860e8218576.png)

@hughsando: Does that tell you anything or do I have to investigate further?

**Comment by hughsando** [comment]
It seems to work on the haxe development version. As far as hxcpp goes, I think this is the more stable version.  Maybe it is a matter of pulling more hxcpp changes over to the branch - not sure how many there are, or if there are many dependencies on the core.

**Comment by hughsando** [comment]
The current gencpp seems entangled with the optional-args representation, so I could not just plonk the new gencpp on top of the old one.

**Comment by Simn** [comment]
I started on the backport here: https://github.com/HaxeFoundation/haxe/commit/22007d3912a692075e05da6a7c1c35321cf45f89

It compiles, but doesn't work yet. We'll have to go through the diff and see what to discard. There's probably a lot of code we can remove because I set `val cppiaAst = false`.

**Comment by sorka95032** [comment]
And just in case we weren't clear, we can forego using function closures on @nativeGen interfaces. Out of a 500K lines of code, i doubt we're using it in more than one location and if there's another, we can fix it to not do that.

**Comment by sorka95032** [comment]
@hughsando, I see you submitted changes to the 3 4 fix  branch a few weeks ago, does the test app compile with this now?

**Comment by Simn** [comment]
I just tested again and the sample here indeed compiles, minus the interface thing.

**Comment by hughsando** [comment]
Yes - think I'm missing the closures only now, but they seemed less important.  Not sure if anything else needs doing.
